source/auth/auth.rst

```diff
@@ -1159,6 +1175,264 @@ If AWS authentication fails for any reason, the cache MUST be cleared.
+The callback mechanism can be used to support both Authentication
+Code Workflows or Device workflows that are not explicitly implemented
+by drivers.  If there is no callback and no DEVICE_NAME, or the
+DEVICE_NAME is set but credentials cannot be automatically obtained,
+the driver MUST raise an error.
+
+
+Supported Device Workflows
+``````````````````````````
+
+Drivers MUST support device workflows for "aws", given
+by the DEVICE_NAME mechanism property.  In all cases the acquired token
+will be given as the ``jwt`` argument and the second client step of the
+OIDC SASL exchange MUST be made directly, skipping the clientStep1.
+Drivers MUST raise an error if both a DEVICE_NAME and username are
+given, since the device workflow will not use the username.
+
+AWS
+___
+
+When the DEVICE_NAME mechanism property is set to "aws", the driver MUST
+attempt to read the value given by the ``AWS_WEB_IDENTITY_TOKEN_FILE`` and
+interpret it as a file path.  The contents of the file are read as the
+access token.
+
+
+Caching Credentials
+```````````````````
+
+Drivers MUST enable caching when callback(s) are provided to the mongo client.
+When an authorization request is made and there is a valid cached response,
+the driver MUST use the cached authorization token if it has not expired.
+
+A cache value is considered valid if it has been created in the past 5 hours.
+The cache is kept alive to preserve the serverStep1 response, as well as
+account for refresh tokens, which typically have an (unknown) lifetime that
+is longer than the access token lifetime.  The refresh token may either be
+stored by the user application, or as part of the request/refresh token
+response in the driver cache. To prevent a memory leak, the driver MUST clear
+invalid cache values at a regular interval, or during every authentication
+attempt.
+
+The cache keys MUST include the username (or empty string) and the
+actually used socket address and port for the current server.
+Using the socket address and port accounts for the case when two different
+servers use the same username but could be configurated differently.
+There is an edge case where if the same username is used and two aliases
+to the same local host address are given, there will be duplicate user/device
+interactions, unless the driver can resolve the local host address as well.
+Note thatbecause we use the server socket address, there will different cache
+keys for each member of a replica set.
+
+The driver MUST cache the serverStep1 reponse as part of the cache value,
+to enable skipping serverStep1 on subsequent authentications of the same
+cache key.
+
+A global cache should be preferred, to prevent multiple browser interactions
+in the case of an authentication code workflow.  However, drivers or dev tools
+can choose to use their own caching scheme if appropriate for their language/
+environment.
+
+A cached value will expire within 5 minutes of the ``expiresInSeconds`` time.
```

Comment:
I assume “within 5 minutes” means “5 minutes _before_ `expiresInSeconds` will be reached”, not “after”, but it might be good to make this wording unambiguous?

What happens if the callback returns an `expiresInSeconds` value that is less than 5 minutes? (If that means not caching the result, then that would be very nice for us in devtools as a way of disabling the cache, but I’m not sure if that’s really what would be most desirable in the big picture.)

Reply:
I clarified the paragraph.